staging/src/kubevirt.io/api/clone/v1alpha1/types.go

```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// VirtualMachineClone is a CRD that clones one VM into another.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +genclient
+// +genclient:nonNamespaced
+type VirtualMachineClone struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   VirtualMachineCloneSpec   `json:"spec" valid:"required"`
+	Status VirtualMachineCloneStatus `json:"status,omitempty"`
+}
+
+type CloneIncludeExclude struct {
+	// +optional
+	CloneAllByDefault *bool `json:"cloneAllByDefault,omitempty"`
+	// +optional
+	// +listType=atomic
+	Exclude []string `json:"exclude,omitempty"`
+	// +optional
+	// +listType=atomic
+	Include []string `json:"include,omitempty"`
+}
+
+type ObjectReference struct {
+	corev1.TypedLocalObjectReference `json:",inline"`
+	Namespace                        string `json:"namespace"`
+}
+
+type VirtualMachineCloneSpec struct {
+	Source *ObjectReference `json:"source"`
+
+	// +optional
+	Target *ObjectReference `json:"target,omitempty"`
+
+	// +optional
+	Annotations *CloneIncludeExclude `json:"annotations,omitempty"`
+	// +optional
+	Labels *CloneIncludeExclude `json:"labels,omitempty"`
+
+	// NewMacAddresses manually sets that target interfaces' mac addresses. The key is the interface name and the
+	// value is the new mac address. If this field is not specified, a new MAC address will
+	// be generated automatically, as for any interface that is not included in this map.
+	// +optional
+	NewMacAddresses map[string]string `json:"newMacAddresses,omitempty"`
+	// NewSMBiosSerial manually sets that target's SMbios serial. If this field is not specified, a new serial will
+	// be generated automatically.
+	// +optional
+	NewSMBiosSerial *string `json:"newSMBiosSerial,omitempty"`
+}
+
+type VirtualMachineClonePhase string
+
+const (
+	PhaseUnset         VirtualMachineClonePhase = ""
+	SnapshotInProgress VirtualMachineClonePhase = "SnapshotInProgress"
+	CreatingTargetVM   VirtualMachineClonePhase = "CreatingTargetVM"
+	RestoreInProgress  VirtualMachineClonePhase = "RestoreInProgress"
+	Succeeded          VirtualMachineClonePhase = "Succeeded"
+	Failed             VirtualMachineClonePhase = "Failed"
+	Unknown            VirtualMachineClonePhase = "Unknown"
+)
+
+type VirtualMachineCloneStatus struct {
+	// +optional
+	// +nullable
+	CreationTime *metav1.Time `json:"creationTime,omitempty"`
+
+	// +optional
+	Phase VirtualMachineClonePhase `json:"phase,omitempty"`
+
+	// +optional
+	// +nullable
+	SnapshotName *string `json:"snapshotName,omitempty"`
+
+	// +optional
+	// +nullable
+	RestoreName *string `json:"restoreName,omitempty"`
+}
```

Comment:
No conditions?  Are snapshot/restorename only set for the duration of the clone operation?

Reply:
Not exactly sure what you mean here. I thought of conditions that might fit the clone object, but didn't come up with anything. There are events such as "Snapshot/Restore is ready", etc.

> Are snapshot/restorename only set for the duration of the clone operation?
Why would they be? Can you elaborate?

Reply:
Conditions are good with `kubectl wait`.  Could see someone creating the clone then waiting for everything to complete.  Events aren't helpful there.

Regarding `restoreName` I assumed that the VMRestore resources were deleted when the restore was complete but maybe not and they are deleted when the clone is deleted via ownerref?